exercises/chapter7/test/Main.purs

```diff
@@ -1,12 +1,103 @@
+        $ Assert.assert "this example should have worked"
+        $ example true true
+    suite "Exercise Group 1" do
+      suite "Use lift2 to write lifted versions of numeric operators" do
+        test "+ (Just)"
+          $ Assert.equal (Just 5)
+          $ (+)
+          <$> (Just 2)
+          <*> (Just 3)
+        test "+ (Nothing on left)"
+          $ Assert.equal Nothing
+          $ (+)
+          <$> Nothing
+          <*> (Just 3)
+        test "+ (Nothing on right)"
+          $ Assert.equal Nothing
+          $ (+)
+          <$> (Just 2)
+          <*> Nothing
+        test "- (Just)"
+          $ Assert.equal (Just (-1))
+          $ (-)
+          <$> (Just 2)
+          <*> (Just 3)
+        test "- (Nothing on left)"
+          $ Assert.equal Nothing
+          $ (-)
+          <$> Nothing
+          <*> (Just 3)
+        test "- (Nothing on right)"
+          $ Assert.equal Nothing
+          $ (-)
+          <$> (Just 2)
+          <*> Nothing
+        test "* (Just)"
+          $ Assert.equal (Just 6)
+          $ (*)
+          <$> (Just 2)
+          <*> (Just 3)
+        test "* (Nothing on left)"
+          $ Assert.equal Nothing
+          $ (*)
+          <$> Nothing
+          <*> (Just 3)
+        test "* (Nothing on right)"
+          $ Assert.equal Nothing
+          $ (*)
+          <$> (Just 2)
+          <*> Nothing
+        test "/ (Just)"
+          $ Assert.equal (Just 2)
+          $ (/)
+          <$> (Just 6)
+          <*> (Just 3)
+        test "/ (Nothing on left)"
+          $ Assert.equal Nothing
+          $ (/)
+          <$> Nothing
+          <*> (Just 3)
+        test "/ (Nothing on right)"
+          $ Assert.equal Nothing
+          $ (/)
+          <$> (Just 2)
+          <*> Nothing
+      suite "Convince yourself that the definition of lift3 type checks" do
+        test "Substituting an Integer type for any of the strings fails to compile"
+          $ Assert.assert "Manually compiled and manually verifed compiling failed"
+          $ true
+      suite "Write a function combineMaybe" do
+        test "Test Maybe of Maybe Int with Just"
+          $ Assert.equal (Just (Just 1))
+          $ combineMaybe (Just $ Just 1)
+        test "Test Maybe of Just with Nothing"
+          $ Assert.equal Nothing
+          $ combineMaybe (Just (Nothing :: Maybe Int))
+        test "Test Nothing"
+          $ Assert.equal (pure Nothing)
+          $ combineMaybe (Nothing :: Maybe (Maybe Int))
+        test "Test List of Maybe Int with Just"
+          $ Assert.equal (Just 1 : Just 2 : Just 3 : Nil)
+          $ combineMaybe (Just $ 1 .. 3)
+        test "Test List of Maybe Int with Nothing"
+          $ Assert.equal (Nothing : Nil)
+          $ combineMaybe (Nothing :: Maybe (List Int))
+        test "Test List of Maybe Int with mix of Just and Nothing"
+          $ Assert.equal (Just (Just 1) : Just Nothing : Just (Just 3) : Nil)
+          $ combineMaybe (Just $ Just 1 : Nothing : Just 3 : Nil)
```

Comment:
I don't know if picking Maybe for our Applicative is the best idea, as far as letting readers easily reason through these tests. Keeping things simple with structures like Array and List might be better.

Reply:
@milesfrain you bring up an interesting idea that didn't occur to me until now:

I was wondering what the payoff of the `combineMaybe` exercise was.  Now I realize that the payoff is `combineMaybe` accomplishes what `combineList` does for `List`, but for all other types of collections as well (`Array`, etc).  It would have helped me to understand the exercise if it had explained that this would be the justification for `combineMaybe`.  Now the regression tests have a reason to do both `List` and `Array` as you imply.

What do you think of adding briefly explaining the advantage of `combineMaybe` over `combineList` in the exercise?  I'd get to that when I implement the changes for the `master` branch.

Reply:
> `combineMaybe` accomplishes what `combineList` does for `List`, but for all other types of collections as well (`Array`, etc).

Not quite. `combineList` also works for `Array` and other Applicatives. In the signatures below, you can substitute `Array` for `f`.
```hs
combineList  :: forall a f. Applicative f => List  (f a) -> f (List  a)
combineMaybe :: forall a f. Applicative f => Maybe (f a) -> f (Maybe a)
```
`sequence` is what accomplishes everything that both `combineMaybe` and `combineList` do. `t` just needs to be a `Traversable`, which both `List` and `Maybe` are.
```hs
combineList  :: forall a f  . Applicative f =>                  List  (f a) -> f (List  a)
combineMaybe :: forall a f  . Applicative f =>                  Maybe (f a) -> f (Maybe a)
sequence     :: forall a f t. Applicative f => Traversable t => t     (f a) -> f (t     a)
```
```hs
listArrayInt :: List (Array Int)
listArrayInt = [ 1, 2 ] : [ 3 ] : Nil

maybeArrayInt :: Maybe (Array Int)
maybeArrayInt = Just [ 1, 2 ]

main :: Effect Unit
main = do
  logShow $ combineList listArrayInt
  logShow $ sequence listArrayInt
  logShow $ combineMaybe maybeArrayInt
  logShow $ sequence maybeArrayInt

{-
Prints:
[(1 : 3 : Nil),(2 : 3 : Nil)]
[(1 : 3 : Nil),(2 : 3 : Nil)]
[(Just 1),(Just 2)]
[(Just 1),(Just 2)]
-}
```

So an alternate solution for the `combineMaybe` exercise is:
```hs
combineMaybe :: forall a f. Applicative f => Maybe (f a) -> f (Maybe a)
combineMaybe = sequence
```
Although I think it's more useful to write a custom version as you have done. Turns out that your answer is identical to [sequence in the Traversable instance for Maybe](https://github.com/purescript/purescript-foldable-traversable/blob/v4.1.1/src/Data/Traversable.purs#L86-L90).

Another interesting point is that if you [search for a function that matches the signature of `combineMaybe` on pursuit](https://pursuit.purescript.org/search?q=forall+a+f.+Applicative+f+%3D%3E+Maybe+%28f+a%29+-%3E+f+%28Maybe+a%29), you find `sequence`.

Reply:
Wow, nice input.  It's interesting that the `sequence` feature comes in the section after this exercise; hence I agree w/ you about keeping the "custom version".

I think the unit tests are sufficient to illustrate the generality of `combineMaybe`.